impl/deps/postgres/instant/instant.go

```diff
@@ -0,0 +1,218 @@
+	if err != nil {
+		return result, err
+	}
+	defer records.Close()
+
+	for records.Next() {
+		var dep deps.Dependency
+		if err := records.Scan(&dep.ID, &dep.Resolved); err != nil {
+			return result, err
+		}
+		result.Dependencies = append(result.Dependencies, dep)
+	}
+
+	return result, nil
+}
+
+// Make generates a new dependency (but it does not store the information about it in the database
+//
+// Generation algorithm: UUID
+func (d *Deps) Make(_ context.Context) (deps.Dependency, error) {
+	id := uuid.New().String()
+
+	return deps.Dependency{
+		ID:       id,
+		Resolved: false,
+	}, nil
+}
+
+// MakeGroup creates a new dependency group
+//
+// NOTE: the new group's id is generated via a UUID algorithm
+func (d *Deps) MakeGroup(ctx context.Context, ids ...string) (string, error) {
+	id := uuid.New().String()
+
+	tx, err := d.client.Begin(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to begin a postgres transaction: %s", tx)
+	}
+	defer tx.Rollback(ctx)
+
+	sql := `INSERT INTO InstantGroups (id, pending)
+			VALUES ($1,
+			        $2 - (SELECT COUNT(*) FROM InstantDependencies d WHERE d.resolved)
+			)`
+	_, err = tx.Exec(ctx, sql, id, len(ids))
+	if err != nil {
+		return "", err
+	}
+
+	for _, dep := range ids {
+		sql := `INSERT INTO InstantGroupDependencies (dependency_id, group_id) VALUES ($1, $2)`
+		_, err := tx.Exec(ctx, sql, dep, id)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if err := tx.Commit(ctx); err != nil {
+		return "", fmt.Errorf("failed to commit the transaction: %s", err)
+	}
+
+	return id, nil
+}
+
+func (d *Deps) Resolve(ctx context.Context, dep string) error {
+	//if err := d.resolvedDeps.Write(ctx, dep); err != nil {
+	//	return fmt.Errorf("failed to write dep to queue: %s", err)
+	//}
+	//return nil
+
+	tx, err := d.client.Begin(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to begin a postgres transaction: %s", tx)
+	}
+	defer tx.Rollback(ctx)
+	// let's goooooo
+	// 1) update dep and check if it has been resolved
+	sql := `
+		WITH previous AS (
+			SELECT resolved
+			FROM InstantDependencies
+			WHERE id = $1
+			LIMIT 1
+		)
+		INSERT INTO InstantDependencies (id, resolved) 
+		VALUES ($1, $2)
+		ON CONFLICT (id) DO UPDATE 
+		SET resolved = $2
+		RETURNING COALESCE((SELECT resolved FROM previous), false) AS previous_resolved;`
+	before := tx.QueryRow(ctx, sql, dep, true)
+	var alreadyResolved bool
+	if err := before.Scan(&alreadyResolved); err != nil {
+		return err
+	}
+	if alreadyResolved {
+		// if it resolved we are done
+		return nil
+	}
+
+	// dep hasn't been resolved, so we need to update groups
+	// we also return groups which have no unresolved deps
+	sql = `
+			UPDATE InstantGroups g
+			SET pending = pending - 1
+			FROM InstantGroupDependencies gd
+			WHERE gd.dependency_id = $1 AND gd.group_id = g.id
+			RETURNING (SELECT g.id WHERE g.pending = 0)` // returns new value
+	resolved, err := tx.Query(ctx, sql, dep)
+	var groups []string
```

Comment:
Despite that `resolved` is automatically closed after a successful scanning in the loop below, we have to `defer resolved.Close()` in order to handle a case when an error happens while scanning.
https://pkg.go.dev/github.com/jackc/pgx#Rows

Reply:
please check if there're any other similar mistakes